File: apps/backend/app/schemas/requirement.py
from datetime import date

from pydantic import BaseModel, Field, field_validator

from app.utils.validators import strip_optional_text


class RequirementCreateSchema(BaseModel):
    category: str = Field(min_length=2, max_length=100)
    subcategory: str | None = Field(default=None, max_length=100)
    number_of_workers: int = Field(ge=1, le=1000)
    work_location: str = Field(min_length=3, max_length=255)
    city: str = Field(min_length=2, max_length=100)
    state: str = Field(min_length=2, max_length=100)
    start_date: date
    duration_days: int = Field(ge=1, le=3650)
    shift_details: str | None = Field(default=None, max_length=255)
    food_required: bool = False
    accommodation_required: bool = False
    budget_amount: int | None = Field(default=None, ge=0)
    notes: str | None = None
    site_latitude: float | None = Field(default=None, ge=-90, le=90)
    site_longitude: float | None = Field(default=None, ge=-180, le=180)
    geofence_radius_meters: int | None = Field(default=2000, ge=0, le=50000)

    @field_validator("notes", "shift_details")
    @classmethod
    def clean_text_fields(cls, value: str | None) -> str | None:
        return strip_optional_text(value)


class RequirementResponseSchema(BaseModel):
    id: int
    category: str
    subcategory: str | None
    number_of_workers: int
    work_location: str
    city: str
    state: str
    start_date: date
    duration_days: int
    shift_details: str | None
    food_required: bool
    accommodation_required: bool
    budget_amount: int | None
    notes: str | None
    status: str
    site_latitude: float | None
    site_longitude: float | None
    geofence_radius_meters: int | None

File: apps/admin/src/types/requirement.ts
export type AdminRequirementListItem = {
  id: number;
  category: string;
  number_of_workers: number;
  city: string;
  status: string;
  start_date: string;
  created_at: string;
  sla_hours: number;
  sla_breach_notified_at: string | null;
};

export type AdminRequirementListData = {
  items: AdminRequirementListItem[];
  page: number;
  page_size: number;
  total: number;
  total_pages: number;
};

export type AdminRequirementListResponse = {
  success: boolean;
  message: string;
  data: AdminRequirementListData;
};

export type AdminRequirementDetail = {
  id: number;
  client_id: number;
  category: string;
  subcategory: string | null;
  number_of_workers: number;
  work_location: string;
  city: string;
  state: string;
  start_date: string;
  duration_days: number;
  shift_details: string | null;
  food_required: boolean;
  accommodation_required: boolean;
  budget_amount: number | null;
  notes: string | null;
  site_latitude: number | null;
  site_longitude: number | null;
  geofence_radius_meters: number | null;
  require_geofence: boolean;
  status: string;
  rejection_reason: string | null;
  cancellation_reason: string | null;
  quote: null | {
    id: number;
    quoted_amount: number;
    rate_per_worker: number | null;
    total_worker_days: number | null;
    advance_amount: number | null;
    payment_model: string;
    valid_until: string | null;
    terms_notes: string | null;
    internal_notes: string | null;
    status: string;
  };
};

export type AdminRequirementDetailResponse = {
  success: boolean;
  message: string;
  data: AdminRequirementDetail;
};

export type RejectRequirementResponse = {
  success: boolean;
  message: string;
  data: {
    id: number;
    status: string;
    rejection_reason: string;
  };
};

export type CancelRequirementResponse = {
  success: boolean;
  message: string;
  data: {
    id: number;
    status: string;
    cancellation_reason: string;
  };
};

export type CreateQuotePayload = {
  requirement_id: number;
  quoted_amount: number;
  rate_per_worker?: number | null;
  total_worker_days?: number | null;
  advance_amount?: number | null;
  payment_model: string;
  valid_until?: string | null;
  terms_notes?: string | null;
  internal_notes?: string | null;
};

export type MarkReviewResponse = {
  success: boolean;
  message: string;
  data: {
    id: number;
    status: string;
  };
};

export type MarkCompleteResponse = {
  success: boolean;
  message: string;
  data: {
    id: number;
    status: string;
  };
};

export type CreateQuoteResponse = {
  success: boolean;
  message: string;
  data: {
    quote_id: number;
    requirement_id: number;
    quote_status: string;
    requirement_status: string;
  };
};

export type RequirementInterestItem = {
  interest_id: number;
  status: string;
  expressed_at: string;
  worker_profile_id: number;
  worker_user_id: number;
  full_name: string;
  city: string;
  category: string;
  subcategory: string | null;
  verification_status: string;
  is_available: boolean;
};

export type RequirementInterestsResponse = {
  success: boolean;
  message: string;
  data: RequirementInterestItem[];
};

export type WorkerPayoutData = {
  id: number;
  amount: number;
  payout_mode: string;
  payout_status: string;
  transaction_reference: string | null;
  notes: string | null;
  paid_at: string | null;
};

export type WorkerPaymentItem = {
  assignment_id: number;
  worker_profile_id: number;
  worker_name: string;
  salary_amount: number | null;
  attendance_days: number;
  half_days: number;
  absent_days: number;
  suggested_amount: number;
  payout: WorkerPayoutData | null;
};

export type WorkerPaymentsResponse = {
  success: boolean;
  message: string;
  data: WorkerPaymentItem[];
};

export type WorkerPaymentRecordPayload = {
  assignment_id: number;
  amount: number;
  payout_mode: string;
  transaction_reference?: string | null;
  notes?: string | null;
};
